### PR TITLE
Add telemetry for retrievers

### PR DESCRIPTION
**Pull request by pmpailis**
This is to update the information present in `ClusterStats` for retrievers. The proposed approach is to introduce a new element named `retriever` under `sections`  that would compute just the top-level usages of a retriever, and additionally a new section named `retrievers` that would contain the names of all the retrievers used anywhere in the tree, i.e. all  top-level usages of a retriever +  any nested retrievers used (e.g. children of `rrf` or `text_similarity_reranker`) 

So assuming that we run a `knn` retriever search, and an `rrf` with a `standard` and a `knn` as direct children, the  suggested output would now be: 
```
{
    "search":
    {
     ...
        "sections":                     
        {
            "retriever": 2             
        },
        "retrievers":                   
        {
            "rrf": 1,
            "knn": 2,
            "standard": 1
        }
    }
}
```

### CODE REVIEW

**Comment by pmpailis** [comment]
@elasticmachine update branch

File: server/src/internalClusterTest/java/org/elasticsearch/search/retriever/RetrieverTelemetryIT.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+    public void testTelemetryForRetrievers() throws IOException {
+
+        if (false == isRetrieverTelemetryEnabled()) {
+            return;
+        }
+
+        // search#1 - this will record 1 entry for "retriever" in `sections`, and 1 for "knn" under `retrievers`
+        {
+            Request request = new Request("GET", INDEX_NAME + "/_search");
```
Comment (carlosdelest) [inline server/src/internalClusterTest/java/org/elasticsearch/search/retriever/RetrieverTelemetryIT.java:81]: Nit - You can probably group these sections in a method like `performQuery(QueryBuilder builder)` to improve readability

File: server/src/main/java/org/elasticsearch/search/retriever/RetrieverBuilder.java [hunk 2]
```diff
@@ -93,7 +93,7 @@ public <T> T namedObject(Class<T> categoryClass, String name, Object context) th
         return parseInnerRetrieverBuilder(parser, context);
     }
 
-    protected static RetrieverBuilder parseInnerRetrieverBuilder(XContentParser parser, RetrieverParserContext context) throws IOException {
+    public static RetrieverBuilder parseInnerRetrieverBuilder(XContentParser parser, RetrieverParserContext context) throws IOException {
```
Comment (carlosdelest) [inline server/src/main/java/org/elasticsearch/search/retriever/RetrieverBuilder.java:96]: Why the change in visibility? 🤔 I can't see it referenced outside this class

File: x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverTelemetryTests.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    public void testTelemetryForRRFRetriever() throws IOException {
+
+        if (false == isRetrieverTelemetryEnabled()) {
+            return;
+        }
+
+        // search#1 - this will record 1 entry for "retriever" in `sections`, and 1 for "knn" under `retrievers`
+        {
```
Comment (carlosdelest) [inline x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverTelemetryTests.java:98]: Ni (Similar to the previous) - refactoring in methods could improve readability

**Review by carlosdelest — APPROVED** [review]
LGTM 👍 

Should we add an `AbstractBWCSerializationTestCase` now that the format has changed for `ClusterStats`?

File: server/src/main/java/org/elasticsearch/search/retriever/RetrieverBuilder.java (hunk 2 above)
Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/search/retriever/RetrieverBuilder.java:96]: No reason 😅  Probably leftover of some initial refactoring/testing

File: server/src/internalClusterTest/java/org/elasticsearch/search/retriever/RetrieverTelemetryIT.java (hunk 1 above)
Comment (pmpailis) [reply server/src/internalClusterTest/java/org/elasticsearch/search/retriever/RetrieverTelemetryIT.java:81]: Yeap, makes sense! Updated.

File: x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverTelemetryTests.java (hunk 3 above)
Comment (pmpailis) [reply x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverTelemetryTests.java:98]: Updated :)

**Comment by pmpailis** [comment]
> Should we add an AbstractBWCSerializationTestCase now that the format has changed for ClusterStats?

The main change lies in how `SearchUsageStats` is structured, which is already covered by a bwc test (`SearchUsageStatsTests#testSerializationBWC`). I've updated it slightly in https://github.com/elastic/elasticsearch/pull/114109/commits/215f120a39b70b67b70ef5b3a523626350794312 to better reflect on the used TransportVersions. 

@carlosdelest  do you think that this would be enough, or should we add a new test class for the top level `ClusterStatsResponse` ?

**Comment by pmpailis** [comment]
@elasticmachine update branch

**Comment by carlosdelest** [comment]
> do you think that this would be enough

@pmpailis , absolutely! Thanks for pointing that out 👍

**Comment by pmpailis** [comment]
@elasticmachine update branch

**Comment by pmpailis** [comment]
run elasticsearch-ci/part-1

**Comment by pmpailis** [comment]
@elasticmachine update branch

**Comment by pmpailis** [comment]
run elasticsearch-ci/packaging-tests-windows-sample

**Pull request merged**